File: VB/WebDAVServer.FileSystemStorage.AspNet/DavHierarchyItem.vb
Imports System
Imports System.Collections.Generic
Imports System.IO
Imports System.Linq
Imports System.Threading.Tasks
Imports ITHit.WebDAV.Server
Imports ITHit.WebDAV.Server.Acl
Imports ITHit.WebDAV.Server.Class2
Imports ITHit.WebDAV.Server.MicrosoftExtensions
Imports WebDAVServer.FileSystemStorage.AspNet.ExtendedAttributes

''' <summary>
''' Base class for WebDAV items (folders, files, etc).
''' </summary>
Public MustInherit Class DavHierarchyItem
    Implements IHierarchyItemAsync, ILockAsync, IMsItemAsync

    ''' <summary>
    ''' Property name to return text anound search phrase.
    ''' </summary>
    Friend Const snippetProperty As String = "snippet"

    ''' <summary>
    ''' Name of properties attribute.
    ''' </summary>
    Friend Const propertiesAttributeName As String = "Properties"

    ''' <summary>
    ''' Name of locks attribute.
    ''' </summary>
    Friend Const locksAttributeName As String = "Locks"

    ''' <summary>
    ''' Gets name of the item.
    ''' </summary>
    Public ReadOnly Property Name As String Implements IHierarchyItemAsync.Name
        Get
            Return fileSystemInfo.Name
        End Get
    End Property

    ''' <summary>
    ''' Gets date when the item was created in UTC.
    ''' </summary>
    Public ReadOnly Property Created As DateTime Implements IHierarchyItemAsync.Created
        Get
            Return fileSystemInfo.CreationTimeUtc
        End Get
    End Property

    ''' <summary>
    ''' Gets date when the item was last modified in UTC.
    ''' </summary>
    Public ReadOnly Property Modified As DateTime Implements IHierarchyItemAsync.Modified
        Get
            Return fileSystemInfo.LastWriteTimeUtc
        End Get
    End Property

    ''' <summary>
    ''' Gets path of the item where each part between slashes is encoded.
    ''' </summary>
    Public Property Path As String Implements IHierarchyItemAsync.Path

    ''' <summary>
    ''' Gets full path for this file/folder in the file system.
    ''' </summary>
    Public ReadOnly Property FullPath As String
        Get
            Return fileSystemInfo.FullName.TrimEnd(System.IO.Path.DirectorySeparatorChar)
        End Get
    End Property

    ''' <summary>
    ''' Corresponding file or folder in the file system.
    ''' </summary>
    Friend fileSystemInfo As FileSystemInfo

    ''' <summary>
    ''' WebDAV Context.
    ''' </summary>
    Protected context As DavContext

    ''' <summary>
    ''' User defined property values.
    ''' </summary>
    Private propertyValues As List(Of PropertyValue)

    ''' <summary>
    ''' Item locks.
    ''' </summary>
    Private locks As List(Of DateLockInfo)

    ''' <summary>
    ''' Initializes a new instance of this class.
    ''' </summary>
    ''' <param name="fileSystemInfo">Corresponding file or folder in the file system.</param>
    ''' <param name="context">WebDAV Context.</param>
    ''' <param name="path">Encoded path relative to WebDAV root folder.</param>
    Protected Sub New(fileSystemInfo As FileSystemInfo, context As DavContext, path As String)
        Me.fileSystemInfo = fileSystemInfo
        Me.context = context
        Me.Path = path
    End Sub

    Public MustOverride Function CopyToAsync(destFolder As IItemCollectionAsync, destName As String, deep As Boolean, multistatus As MultistatusException) As Task Implements IHierarchyItemAsync.CopyToAsync

    Public MustOverride Function MoveToAsync(destFolder As IItemCollectionAsync, destName As String, multistatus As MultistatusException) As Task Implements IHierarchyItemAsync.MoveToAsync

    Public MustOverride Function DeleteAsync(multistatus As MultistatusException) As Task Implements IHierarchyItemAsync.DeleteAsync

    Public Async Function GetPropertiesAsync(props As IList(Of PropertyName), allprop As Boolean) As Task(Of IEnumerable(Of PropertyValue)) Implements IHierarchyItemAsync.GetPropertiesAsync
        Dim propertyValues As List(Of PropertyValue) = Await GetPropertyValuesAsync()
        Dim snippet As PropertyName = props.FirstOrDefault(Function(s) s.Name = snippetProperty)
        If snippet.Name = snippetProperty AndAlso TypeOf Me Is DavFile Then
            propertyValues.Add(New PropertyValue(snippet, CType(Me, DavFile).Snippet))
        End If

        If Not allprop Then
            propertyValues = propertyValues.Where(Function(p) props.Contains(p.QualifiedName)).ToList()
        End If

        Return propertyValues
    End Function

    Public Async Function GetPropertyNamesAsync() As Task(Of IEnumerable(Of PropertyName)) Implements IHierarchyItemAsync.GetPropertyNamesAsync
        Dim propertyValues As IList(Of PropertyValue) = Await GetPropertyValuesAsync()
        Return propertyValues.Select(Function(p) p.QualifiedName)
    End Function

    Private Async Function GetPropertyValuesAsync() As Task(Of List(Of PropertyValue))
        If propertyValues Is Nothing Then
            propertyValues = Await fileSystemInfo.GetExtendedAttributeAsync(Of List(Of PropertyValue))(propertiesAttributeName)
        End If

        Return propertyValues
    End Function

    Public Async Function UpdatePropertiesAsync(setProps As IList(Of PropertyValue), delProps As IList(Of PropertyName), multistatus As MultistatusException) As Task Implements IHierarchyItemAsync.UpdatePropertiesAsync
        Await RequireHasTokenAsync()
        Dim propertyValues As List(Of PropertyValue) = Await GetPropertyValuesAsync()
        For Each propToSet As PropertyValue In setProps
            If propToSet.QualifiedName.Namespace = "urn:schemas-microsoft-com:" Then
                Select Case propToSet.QualifiedName.Name
                    Case "Win32CreationTime"
                        fileSystemInfo.CreationTimeUtc = DateTime.Parse(propToSet.Value,
                                                                       New System.Globalization.CultureInfo("en-US")).ToUniversalTime()
                    Case "Win32LastModifiedTime"
                        fileSystemInfo.LastWriteTimeUtc = DateTime.Parse(propToSet.Value,
                                                                        New System.Globalization.CultureInfo("en-US")).ToUniversalTime()
                    Case Else
                        context.Logger.LogDebug(String.Format("Unspecified case: DavHierarchyItem.UpdateProperties {0} from {1} namesapce",
                                                             propToSet.QualifiedName.Name, propToSet.QualifiedName.Namespace))
                End Select
            Else
                Dim existingProp As PropertyValue = propertyValues.FirstOrDefault(Function(p) p.QualifiedName = propToSet.QualifiedName)
                If existingProp IsNot Nothing Then
                    existingProp.Value = propToSet.Value
                Else
                    propertyValues.Add(propToSet)
                End If
            End If
        Next

        propertyValues.RemoveAll(Function(prop) delProps.Contains(prop.QualifiedName))
        Await fileSystemInfo.SetExtendedAttributeAsync(propertiesAttributeName, propertyValues)
        Await context.socketService.NotifyRefreshAsync(GetParentPath(Path))
    End Function

    Public Async Function GetFileAttributesAsync() As Task(Of FileAttributes) Implements IMsItemAsync.GetFileAttributesAsync
        If Name.StartsWith(".") Then
            Return fileSystemInfo.Attributes Or FileAttributes.Hidden
        End If

        Return fileSystemInfo.Attributes
    End Function

    Public Async Function SetFileAttributesAsync(value As FileAttributes) As Task Implements IMsItemAsync.SetFileAttributesAsync
        File.SetAttributes(fileSystemInfo.FullName, value)
    End Function

    Public Async Function GetActiveLocksAsync() As Task(Of IEnumerable(Of LockInfo)) Implements ILockAsync.GetActiveLocksAsync
        Dim locks As List(Of DateLockInfo) = Await GetLocksAsync()
        If locks Is Nothing Then
            Return New List(Of LockInfo)()
        End If

        Dim lockInfoList As IEnumerable(Of LockInfo) = locks.Select(Function(l) New LockInfo With {.IsDeep = l.IsDeep, .Level = l.Level, .Owner = l.ClientOwner, .LockRoot = l.LockRoot, .TimeOut = If(l.Expiration = DateTime.MaxValue, TimeSpan.MaxValue, l.Expiration - DateTime.UtcNow), .Token = l.LockToken}).ToList()
        Return lockInfoList
    End Function

    Public Async Function LockAsync(level As LockLevel, isDeep As Boolean, requestedTimeOut As TimeSpan?, owner As String) As Task(Of LockResult) Implements ILockAsync.LockAsync
        Await RequireUnlockedAsync(level = LockLevel.Shared)
        Dim token As String = Guid.NewGuid().ToString()
        Dim timeOut As TimeSpan = TimeSpan.FromMinutes(5)
        If requestedTimeOut.HasValue AndAlso requestedTimeOut < TimeSpan.MaxValue Then
            timeOut = requestedTimeOut.Value
        End If

        Dim lockInfo As DateLockInfo = New DateLockInfo With {.Expiration = DateTime.UtcNow + timeOut, .IsDeep = False, .Level = level, .LockRoot = Path, .LockToken = token, .ClientOwner = owner, .TimeOut = timeOut}
        Await SaveLockAsync(lockInfo)
        Await context.socketService.NotifyRefreshAsync(GetParentPath(Path))
        Return New LockResult(lockInfo.LockToken, lockInfo.TimeOut)
    End Function

    Public Async Function RefreshLockAsync(token As String, requestedTimeOut As TimeSpan?) As Task(Of RefreshLockResult) Implements ILockAsync.RefreshLockAsync
        If String.IsNullOrEmpty(token) Then
            Throw New DavException("Lock can not be found.", DavStatus.BAD_REQUEST)
        End If

        Dim locks As List(Of DateLockInfo) = Await GetLocksAsync(getAllWithExpired:=True)
        Dim lockInfo As DateLockInfo = locks.SingleOrDefault(Function(x) x.LockToken = token)
        If lockInfo Is Nothing OrElse lockInfo.Expiration <= DateTime.UtcNow Then
            Throw New DavException("Lock can not be found.", DavStatus.CONFLICT)
        Else
            lockInfo.TimeOut = TimeSpan.FromMinutes(5)
            If requestedTimeOut.HasValue AndAlso requestedTimeOut < TimeSpan.MaxValue Then
                lockInfo.TimeOut = requestedTimeOut.Value
            End If

            lockInfo.Expiration = DateTime.UtcNow + lockInfo.TimeOut
            Await SaveLockAsync(lockInfo)
        End If

        Await context.socketService.NotifyRefreshAsync(GetParentPath(Path))
        Return New RefreshLockResult(lockInfo.Level, lockInfo.IsDeep, lockInfo.TimeOut, lockInfo.ClientOwner)
    End Function

    Public Async Function UnlockAsync(lockToken As String) As Task Implements ILockAsync.UnlockAsync
        If String.IsNullOrEmpty(lockToken) Then
            Throw New DavException("Lock can not be found.", DavStatus.BAD_REQUEST)
        End If

        Dim locks As List(Of DateLockInfo) = Await GetLocksAsync(getAllWithExpired:=True)
        Dim lockInfo As DateLockInfo = locks.SingleOrDefault(Function(x) x.LockToken = lockToken)
        Await RemoveExpiredLocksAsync(lockToken)
        If lockInfo Is Nothing OrElse lockInfo.Expiration <= DateTime.UtcNow Then
            Throw New DavException("The lock could not be found.", DavStatus.CONFLICT)
        End If

        Await context.socketService.NotifyRefreshAsync(GetParentPath(Path))
    End Function

    Public Async Function RequireHasTokenAsync(Optional skipShared As Boolean = False) As Task
        Dim locks As List(Of DateLockInfo) = Await GetLocksAsync()
        If locks IsNot Nothing AndAlso locks.Any() Then
            Dim clientLockTokens As IList(Of String) = context.Request.ClientLockTokens
            If locks.All(Function(l) Not clientLockTokens.Contains(l.LockToken)) Then
                Throw New LockedException()
            End If
        End If
    End Function

    Public Async Function RequireUnlockedAsync(skipShared As Boolean) As Task
        Dim locks As List(Of DateLockInfo) = Await GetLocksAsync()
        If locks IsNot Nothing AndAlso locks.Any() Then
            If(skipShared AndAlso locks.Any(Function(l) l.Level = LockLevel.Exclusive)) OrElse (Not skipShared AndAlso locks.Any()) Then
                Throw New LockedException()
            End If
        End If
    End Function

    Private Async Function GetLocksAsync(Optional getAllWithExpired As Boolean = False) As Task(Of List(Of DateLockInfo))
        If locks Is Nothing Then
            locks = Await fileSystemInfo.GetExtendedAttributeAsync(Of List(Of DateLockInfo))(locksAttributeName)
            If locks IsNot Nothing Then
                locks.ForEach(Sub(l) __InlineAssignHelper(l.LockRoot, Path))
            End If
        End If

        If locks Is Nothing Then
            Return New List(Of DateLockInfo)()
        End If

        If getAllWithExpired Then
            Return locks
        Else
            Return locks.Where(Function(x) x.Expiration > DateTime.UtcNow).ToList()
        End If
    End Function

    Private Async Function SaveLockAsync(lockInfo As DateLockInfo) As Task
        Dim locks As List(Of DateLockInfo) = Await GetLocksAsync(getAllWithExpired:=True)
        'remove all expired locks
        'await RemoveExpiretLocksAsync();
        'you can call this method but it will be second file operation
        locks.RemoveAll(Function(x) x.Expiration <= DateTime.UtcNow)
        If locks.Any(Function(x) x.LockToken = lockInfo.LockToken) Then
            Dim existingLock As DateLockInfo = locks.[Single](Function(x) x.LockToken = lockInfo.LockToken)
            existingLock.TimeOut = lockInfo.TimeOut
            existingLock.Level = lockInfo.Level
            existingLock.IsDeep = lockInfo.IsDeep
            existingLock.LockRoot = lockInfo.LockRoot
            existingLock.Expiration = lockInfo.Expiration
            existingLock.ClientOwner = lockInfo.ClientOwner
        Else
            'add new item
            locks.Add(lockInfo)
        End If

        Await fileSystemInfo.SetExtendedAttributeAsync(locksAttributeName, locks)
    End Function

    Private Async Function RemoveExpiredLocksAsync(Optional unlockedToken As String = Nothing) As Task
        Dim locks As List(Of DateLockInfo) = Await GetLocksAsync(getAllWithExpired:=True)
        'remove expired and current lock
        locks.RemoveAll(Function(x) x.Expiration <= DateTime.UtcNow)
        If Not String.IsNullOrEmpty(unlockedToken) Then
            locks.RemoveAll(Function(x) x.LockToken = unlockedToken)
        End If

        Await fileSystemInfo.SetExtendedAttributeAsync(locksAttributeName, locks)
    End Function

    Protected Shared Function GetParentPath(path As String) As String
        Dim parentPath As String = $"/{path.Trim("/"c)}"
        Dim index As Integer = parentPath.LastIndexOf("/")
        parentPath = parentPath.Substring(0, index)
        Return parentPath
    End Function

    <Obsolete("Please refactor code that uses this function, it is a simple work-around to simulate inline assignment in VB!")>
    Private Shared Function __InlineAssignHelper(Of T)(ByRef target As T, value As T) As T
        target = value
        Return value
    End Function
End Class

File: CS/WebDAVServer.FileSystemSynchronization.AspNetCore/DavHierarchyItem.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using System.Threading.Tasks;

using ITHit.WebDAV.Server;
using ITHit.WebDAV.Server.Acl;
using IPrincipal = ITHit.WebDAV.Server.Acl.IPrincipal;
using ITHit.WebDAV.Server.Extensibility;
using ITHit.WebDAV.Server.Class2;
using ITHit.WebDAV.Server.MicrosoftExtensions;
using ITHit.WebDAV.Server.Synchronization;
using WebDAVServer.FileSystemSynchronization.AspNetCore.ExtendedAttributes;

namespace WebDAVServer.FileSystemSynchronization.AspNetCore
{
    /// <summary>
    /// Base class for WebDAV items (folders, files, etc).
    /// </summary>
    public abstract class DavHierarchyItem : IHierarchyItem, ILock, IMsItem, IBind, IChangedItem
    {
        /// <summary>
        /// Property name to return text anound search phrase.
        /// </summary>
        internal const string snippetProperty = "snippet";
        /// <summary>
        /// Property name of metadata Etag.
        /// </summary>
        internal const string metadataEtagProperty = "metadata-Etag";

        /// <summary>
        /// Name of properties attribute.
        /// </summary>
        internal const string propertiesAttributeName = "Properties";

        /// <summary>
        /// Name of locks attribute.
        /// </summary>
        internal const string locksAttributeName = "Locks";

        /// <summary>
        /// Gets name of the item.
        /// </summary>
        public string Name { get { return fileSystemInfo.Name; } }

        /// <summary>
        /// Gets date when the item was created in UTC.
        /// </summary>
        public DateTime Created { get { return fileSystemInfo.CreationTimeUtc; } }

        /// <summary>
        /// Gets date when the item was last modified in UTC.
        /// </summary>
        public DateTime Modified { get { return fileSystemInfo.LastWriteTimeUtc; } }

        /// <summary>
        /// Gets path of the item where each part between slashes is encoded.
        /// </summary>
        public string Path { get; private set; }

        /// <summary>
        /// Gets full path for this file/folder in the file system.
        /// </summary>
        public string FullPath { get { return fileSystemInfo.FullName.TrimEnd(System.IO.Path.DirectorySeparatorChar); } }
        public Change ChangeType
        {
            get => fileSystemInfo.Attributes.HasFlag(FileAttributes.Hidden) &&
                (this is DavFolder || (this as DavFile).ContentLength == 0) ? Change.Deleted : Change.Changed;
        }

        /// <summary>
        /// Unique identifier of the resource.
        /// </summary>
        public string Id { get => $"{context.Request.UrlPrefix}/ID/{fileSystemInfo.GetId()}"; }

        /// <summary>
        /// Unique identifier of the resource parent. 
        /// </summary>
        public string ParentId { get => $"{context.Request.UrlPrefix}/ID/{Directory.GetParent(fileSystemInfo.FullName).GetId()}"; }

        /// <summary>
        /// Corresponding file or folder in the file system.
        /// </summary>
        internal FileSystemInfo fileSystemInfo;

        /// <summary>
        /// WebDAV Context.
        /// </summary>
        protected DavContext context;

        /// <summary>
        /// Initializes a new instance of this class.
        /// </summary>
        /// <param name="fileSystemInfo">Corresponding file or folder in the file system.</param>
        /// <param name="context">WebDAV Context.</param>
        /// <param name="path">Encoded path relative to WebDAV root folder.</param>
        protected DavHierarchyItem(FileSystemInfo fileSystemInfo, DavContext context, string path)
        {
            this.fileSystemInfo = fileSystemInfo;
            this.context = context;
            this.Path = path;
        }

        /// <summary>
        /// Creates a copy of this item with a new name in the destination folder.
        /// </summary>
        /// <param name="destFolder">Destination folder.</param>
        /// <param name="destName">Name of the destination item.</param>
        /// <param name="deep">Indicates whether to copy entire subtree.</param>
        /// <param name="multistatus">If some items fail to copy but operation in whole shall be continued, add
        /// information about the error into <paramref name="multistatus"/> using 
        /// <see cref="MultistatusException.AddInnerException(string,ITHit.WebDAV.Server.DavException)"/>.
        /// </param>
        public abstract Task CopyToAsync(IItemCollection destFolder, string destName, bool deep, MultistatusException multistatus);

        /// <summary>
        /// Creates a copy of this item with a new name in the destination folder.
        /// </summary>
        /// <param name="destFolder">Destination folder.</param>
        /// <param name="destName">Name of the destination item.</param>
        /// <param name="deep">Indicates whether to copy entire subtree.</param>
        /// <param name="multistatus">If some items fail to copy but operation in whole shall be continued, add
        /// information about the error into <paramref name="multistatus"/> using 
        /// <see cref="MultistatusException.AddInnerException(string,ITHit.WebDAV.Server.DavException)"/>.
        /// </param>
        /// <param name="recursionDepth">Recursion depth.</param>
        public abstract Task CopyToInternalAsync(IItemCollection destFolder, string destName, bool deep, MultistatusException multistatus, int recursionDepth);

        /// <summary>
        /// Moves this item to the destination folder under a new name.
        /// </summary>
        /// <param name="destFolder">Destination folder.</param>
        /// <param name="destName">Name of the destination item.</param>
        /// <param name="multistatus">If some items fail to copy but operation in whole shall be continued, add
        /// information about the error into <paramref name="multistatus"/> using 
        /// <see cref="MultistatusException.AddInnerException(string,ITHit.WebDAV.Server.DavException)"/>.
        /// </param>
        public abstract Task MoveToAsync(IItemCollection destFolder, string destName, MultistatusException multistatus);

        /// <summary>
        /// Moves this item to the destination folder under a new name.
        /// </summary>
        /// <param name="destFolder">Destination folder.</param>
        /// <param name="destName">Name of the destination item.</param>
        /// <param name="multistatus">If some items fail to copy but operation in whole shall be continued, add
        /// information about the error into <paramref name="multistatus"/> using 
        /// <see cref="MultistatusException.AddInnerException(string,ITHit.WebDAV.Server.DavException)"/>.
        /// </param>
        /// <param name="recursionDepth">Recursion depth.</param>
        public abstract Task MoveToInternalAsync(IItemCollection destFolder, string destName, MultistatusException multistatus, int recursionDepth);

        /// <summary>
        /// Deletes this item.
        /// </summary>
        /// <param name="multistatus">If some items fail to delete but operation in whole shall be continued, add
        /// information about the error into <paramref name="multistatus"/> using
        /// <see cref="MultistatusException.AddInnerException(string,ITHit.WebDAV.Server.DavException)"/>.
        /// </param>
        public abstract Task DeleteAsync(MultistatusException multistatus);

        /// <summary>
        /// Deletes this item.
        /// </summary>
        /// <param name="multistatus">If some items fail to delete but operation in whole shall be continued, add
        /// information about the error into <paramref name="multistatus"/> using
        /// <see cref="MultistatusException.AddInnerException(string,ITHit.WebDAV.Server.DavException)"/>.
        /// </param>
        /// <param name="recursionDepth">Recursion depth.</param>
        public abstract Task DeleteInternalAsync(MultistatusException multistatus, int recursionDepth);

        /// <summary>
        /// Retrieves user defined property values.
        /// </summary>
        /// <param name="names">Names of dead properties which values to retrieve.</param>
        /// <param name="allprop">Whether all properties shall be retrieved.</param>
        /// <returns>Property values.</returns>
        public async Task<IEnumerable<PropertyValue>> GetPropertiesAsync(IList<PropertyName> props, bool allprop)
        {
            List<PropertyValue> propertyValues = await GetPropertyValuesAsync();

            PropertyName snippet = props.FirstOrDefault(s => s.Name == snippetProperty);
            if (snippet.Name == snippetProperty && this is DavFile)
            {
                propertyValues.Add(new PropertyValue(snippet, ((DavFile)this).Snippet));
            }
            PropertyName metadataEtag = props.FirstOrDefault(s => s.Name == metadataEtagProperty);
            if (metadataEtag.Name == metadataEtagProperty)
            {
                propertyValues.Add(new PropertyValue(metadataEtag,
                    ((await fileSystemInfo.GetExtendedAttributeAsync<int?>("MetadateSerialNumber")) ?? 0).ToString()));
            }
            if (!allprop)
            {
                propertyValues = propertyValues.Where(p => props.Contains(p.QualifiedName)).ToList();
            }

            return propertyValues;
        }

        /// <summary>
        /// Retrieves names of all user defined properties.
        /// </summary>
        /// <returns>Property names.</returns>
        public async Task<IEnumerable<PropertyName>> GetPropertyNamesAsync()
        {
            IList<PropertyValue> propertyValues = await GetPropertyValuesAsync();
            return propertyValues.Select(p => p.QualifiedName);
        }

        /// <summary>
        /// Retrieves list of user defined propeties for this item.
        /// </summary>
        /// <returns>List of user defined properties.</returns>
        private async Task<List<PropertyValue>> GetPropertyValuesAsync()
        {
            List<PropertyValue> properties = new List<PropertyValue>();
            if (await fileSystemInfo.HasExtendedAttributeAsync(propertiesAttributeName))
            {
                properties = await fileSystemInfo.GetExtendedAttributeAsync<List<PropertyValue>>(propertiesAttributeName);
            }

            return properties;
        }

        /// <summary>
        /// Saves property values to extended attribute.
        /// </summary>
        /// <param name="setProps">Properties to be set.</param>
        /// <param name="delProps">Properties to be deleted.</param>
        /// <param name="multistatus">Information about properties that failed to create, update or delate.</param>
        public async Task UpdatePropertiesAsync(IList<PropertyValue> setProps, IList<PropertyName> delProps, MultistatusException multistatus)
        {
            await RequireHasTokenAsync();
            List<PropertyValue> propertyValues = await GetPropertyValuesAsync();
            foreach (PropertyValue propToSet in setProps)
            {
                // Microsoft Mini-redirector may update file creation date, modification date and access time passing properties:
                // <Win32CreationTime xmlns="urn:schemas-microsoft-com:">Thu, 28 Mar 2013 20:15:34 GMT</Win32CreationTime>
                // <Win32LastModifiedTime xmlns="urn:schemas-microsoft-com:">Thu, 28 Mar 2013 20:36:24 GMT</Win32LastModifiedTime>
                // <Win32LastAccessTime xmlns="urn:schemas-microsoft-com:">Thu, 28 Mar 2013 20:36:24 GMT</Win32LastAccessTime>
                // In this case update creation and modified date in your storage or do not save this properties at all, otherwise 
                // Windows Explorer will display creation and modification date from this props and it will differ from the values 
                // in the Created and Modified fields in your storage 
                if (propToSet.QualifiedName.Namespace == "urn:schemas-microsoft-com:")
                {
                    switch (propToSet.QualifiedName.Name)
                    {
                        case "Win32CreationTime":
                            fileSystemInfo.CreationTimeUtc = DateTime.Parse(propToSet.Value,
                                new System.Globalization.CultureInfo("en-US")).ToUniversalTime();
                            break;
                        case "Win32LastModifiedTime":
                            fileSystemInfo.LastWriteTimeUtc = DateTime.Parse(propToSet.Value,
                                new System.Globalization.CultureInfo("en-US")).ToUniversalTime();
                            break;
                        default:
                            context.Logger.LogDebug(string.Format("Unspecified case: DavHierarchyItem.UpdateProperties {0} from {1} namesapce",
                                propToSet.QualifiedName.Name, propToSet.QualifiedName.Namespace));
                            break;
                    }
                }
                else
                {
                    PropertyValue existingProp = propertyValues.FirstOrDefault(p => p.QualifiedName == propToSet.QualifiedName);

                    if (existingProp != null)
                    {
                        existingProp.Value = propToSet.Value;
                    }
                    else
                    {
                        propertyValues.Add(propToSet);
                    }
                }
            }

            propertyValues.RemoveAll(prop => delProps.Contains(prop.QualifiedName));

            await fileSystemInfo.SetExtendedAttributeAsync(propertiesAttributeName, propertyValues);
            await UpdateMetadateEtagAsync();
            await context.socketService.NotifyUpdatedAsync(Path, GetWebSocketID());
        }

        /// <summary>
        /// Returns Windows file attributes (readonly, hidden etc.) for this file/folder.
        /// </summary>
        /// <returns>Windows file attributes.</returns>
        public async Task<FileAttributes> GetFileAttributesAsync()
        {
            if (Name.StartsWith("."))
            {
                return fileSystemInfo.Attributes | FileAttributes.Hidden;
            }
            return fileSystemInfo.Attributes;
        }

        /// <summary>
        /// Sets Windows file attributes (readonly, hidden etc.) on this item.
        /// </summary>
        /// <param name="value">File attributes.</param>
        public async Task SetFileAttributesAsync(FileAttributes value)
        {
            File.SetAttributes(fileSystemInfo.FullName, value);
        }

        /// <summary>
        /// Retrieves non expired locks for this item.
        /// </summary>
        /// <returns>Enumerable with information about locks.</returns>
        public async Task<IEnumerable<LockInfo>> GetActiveLocksAsync()
        {
            List<DateLockInfo> locks = await GetLocksAsync();
            if (locks == null)
            {
                return new List<LockInfo>();
            }

            IEnumerable<LockInfo> lockInfoList = locks.Select(l => new LockInfo
            {
                IsDeep = l.IsDeep,
                Level = l.Level,
                Owner = l.ClientOwner,
                LockRoot = l.LockRoot,
                TimeOut = l.Expiration == DateTime.MaxValue ?
                            TimeSpan.MaxValue :
                            l.Expiration - DateTime.UtcNow,
                Token = l.LockToken
            }).ToList();

            return lockInfoList;
        }

        /// <summary>
        /// Locks this item.
        /// </summary>
        /// <param name="level">Whether lock is share or exclusive.</param>
        /// <param name="isDeep">Whether lock is deep.</param>
        /// <param name="requestedTimeOut">Lock timeout which was requested by client.
        /// Server may ignore this parameter and set any timeout.</param>
        /// <param name="owner">Owner of the lock as specified by client.</param> 
        /// <returns>
        /// Instance of <see cref="LockResult"/> with information about the lock.
        /// </returns>
        public async Task<LockResult> LockAsync(LockLevel level, bool isDeep, TimeSpan? requestedTimeOut, string owner)
        {
            await RequireUnlockedAsync(level == LockLevel.Shared);
            string token = Guid.NewGuid().ToString();

            // If timeout is absent or infinit timeout requested,
            // grant 5 minute lock.
            TimeSpan timeOut = TimeSpan.FromMinutes(5);

            if (requestedTimeOut.HasValue && requestedTimeOut < TimeSpan.MaxValue)
            {
                timeOut = requestedTimeOut.Value;
            }

            DateLockInfo lockInfo = new DateLockInfo
            {
                Expiration = DateTime.UtcNow + timeOut,
                IsDeep = false,
                Level = level,
                LockRoot = Path,
                LockToken = token,
                ClientOwner = owner,
                TimeOut = timeOut
            };

            await SaveLockAsync(lockInfo);
            await context.socketService.NotifyLockedAsync(Path, GetWebSocketID());

            return new LockResult(lockInfo.LockToken, lockInfo.TimeOut);
        }

        /// <summary>
        /// Updates lock timeout information on this item.
        /// </summary>
        /// <param name="token">Lock token.</param>
        /// <param name="requestedTimeOut">Lock timeout which was requested by client.
        /// Server may ignore this parameter and set any timeout.</param>
        /// <returns>
        /// Instance of <see cref="LockResult"/> with information about the lock.
        /// </returns>
        public async Task<RefreshLockResult> RefreshLockAsync(string token, TimeSpan? requestedTimeOut)
        {
            if (string.IsNullOrEmpty(token))
            {
                throw new DavException("Lock can not be found.", DavStatus.BAD_REQUEST);
            }

            List<DateLockInfo> locks = await GetLocksAsync(getAllWithExpired: true);
            DateLockInfo lockInfo = locks.SingleOrDefault(x => x.LockToken == token);

            if (lockInfo == null || lockInfo.Expiration <= DateTime.UtcNow)
            {
                throw new DavException("Lock can not be found.", DavStatus.CONFLICT);
            }
            else
            {
                lockInfo.TimeOut = TimeSpan.FromMinutes(5);

                if (requestedTimeOut.HasValue && requestedTimeOut < TimeSpan.MaxValue)
                {
                    lockInfo.TimeOut = requestedTimeOut.Value;
                }


                lockInfo.Expiration = DateTime.UtcNow + lockInfo.TimeOut;

                await SaveLockAsync(lockInfo);
            }
            await context.socketService.NotifyLockedAsync(Path, GetWebSocketID());

            return new RefreshLockResult(lockInfo.Level, lockInfo.IsDeep, lockInfo.TimeOut, lockInfo.ClientOwner);
        }

        /// <summary>
        /// Removes lock with the specified token from this item.
        /// </summary>
        /// <param name="lockToken">Lock with this token should be removed from the item.</param>
        public async Task UnlockAsync(string lockToken)
        {
            if (string.IsNullOrEmpty(lockToken))
            {
                throw new DavException("Lock can not be found.", DavStatus.BAD_REQUEST);
            }

            List<DateLockInfo> locks = await GetLocksAsync(getAllWithExpired: true);
            DateLockInfo lockInfo = locks.SingleOrDefault(x => x.LockToken == lockToken);

            await RemoveExpiredLocksAsync(lockToken);

            if (lockInfo == null || lockInfo.Expiration <= DateTime.UtcNow)
            {
                throw new DavException("The lock could not be found.", DavStatus.CONFLICT);
            }
            await context.socketService.NotifyUnLockedAsync(Path, GetWebSocketID());
        }

        /// <summary>
        /// Check that if the item is locked then client has submitted correct lock token.
        /// </summary>
        public async Task RequireHasTokenAsync(bool skipShared = false)
        {
            List<DateLockInfo> locks = await GetLocksAsync();
            if (locks != null && locks.Any())
            {
                IList<string> clientLockTokens = context.Request.GetClientLockTokens();
                if (locks.All(l => !clientLockTokens.Contains(l.LockToken)))
                {
                    throw new LockedException();
                }
            }
        }

        /// <summary>
        /// Ensure that there are no active locks on the item.
        /// </summary>
        /// <param name="skipShared">Whether shared locks shall be checked.</param>
        public async Task RequireUnlockedAsync(bool skipShared)
        {
            List<DateLockInfo> locks = await GetLocksAsync();

            if (locks != null && locks.Any())
            {
                if ((skipShared && locks.Any(l => l.Level == LockLevel.Exclusive))
                    || (!skipShared && locks.Any()))
                {
                    throw new LockedException();
                }
            }
        }

        /// <summary>
        /// Retrieves non-expired locks acquired on this item.
        /// </summary>
        /// <param name="getAllWithExpired">Indicate needed return expired locks.</param>
        /// <returns>List of locks with their expiration dates.</returns>
        private async Task<List<DateLockInfo>> GetLocksAsync(bool getAllWithExpired = false)
        {

            List<DateLockInfo> locks = new List<DateLockInfo>();

            if (await fileSystemInfo.HasExtendedAttributeAsync(locksAttributeName))
            {
                locks = await fileSystemInfo.GetExtendedAttributeAsync<List<DateLockInfo>>(locksAttributeName);

                if (locks != null)
                {
                    locks.ForEach(l => l.LockRoot = Path);
                }
            }

            if (getAllWithExpired)
            {
                return locks;
            }
            else
            {
                return locks.Where(x => x.Expiration > DateTime.UtcNow).ToList();
            }
        }

        /// <summary>
        /// Saves lock acquired on this file/folder.
        /// </summary>
        private async Task SaveLockAsync(DateLockInfo lockInfo)
        {
            List<DateLockInfo> locks = await GetLocksAsync(getAllWithExpired: true);

            //remove all expired locks
            //await RemoveExpiretLocksAsync();
            //you can call this method but it will be second file operation
            locks.RemoveAll(x => x.Expiration <= DateTime.UtcNow);

            if (locks.Any(x => x.LockToken == lockInfo.LockToken))
            {
                //update value
                DateLockInfo existingLock = locks.Single(x => x.LockToken == lockInfo.LockToken);
                existingLock.TimeOut = lockInfo.TimeOut;

                existingLock.Level = lockInfo.Level;
                existingLock.IsDeep = lockInfo.IsDeep;
                existingLock.LockRoot = lockInfo.LockRoot;
                existingLock.Expiration = lockInfo.Expiration;
                existingLock.ClientOwner = lockInfo.ClientOwner;
            }
            else
            {
                //add new item
                locks.Add(lockInfo);
            }

            await fileSystemInfo.SetExtendedAttributeAsync(locksAttributeName, locks);
            await UpdateMetadateEtagAsync();
        }

        private async Task RemoveExpiredLocksAsync(string unlockedToken)
        {
            List<DateLockInfo> locks = await GetLocksAsync(getAllWithExpired: true);

            //remove expired and current lock
            locks.RemoveAll(x => x.Expiration <= DateTime.UtcNow);

            //remove from token
            if (!string.IsNullOrEmpty(unlockedToken))
            {
                locks.RemoveAll(x => x.LockToken == unlockedToken);
            }

            await fileSystemInfo.SetExtendedAttributeAsync(locksAttributeName, locks);
            await UpdateMetadateEtagAsync();
        }

        /// <summary>
        /// Gets element's parent path. 
        /// </summary>
        /// <param name="path">Element's path.</param>
        /// <returns>Path to parent element.</returns>
        protected static string GetParentPath(string path)
        {
            string parentPath = string.Format("/{0}", path.Trim('/'));
            int index = parentPath.LastIndexOf("/");
            parentPath = parentPath.Substring(0, index);
            return parentPath;
        }

        /// <summary>
        /// Returns WebSocket client ID.
        /// </summary>
        /// <returns>Client ID.</returns>
        protected string GetWebSocketID()
        {
            return context.Request.Headers.ContainsKey("InstanceId") ?
                context.Request.Headers["InstanceId"] : string.Empty;
        }
        protected async Task UpdateMetadateEtagAsync()
        {
            int serialNumber = await fileSystemInfo.GetExtendedAttributeAsync<int?>("MetadateSerialNumber") ?? 0;

            await fileSystemInfo.SetExtendedAttributeAsync("MetadateSerialNumber", ++serialNumber);
        }
    }
}
